[0x4F776C] iP

0x4F776C opened the pull request:
# **TrashBot** 🤖

> "Trash your tasks, not your productivity!" 

TrashBot is your command-line companion for efficient task management. It's:

* Simple to use
* Command-line based
* ~~BASIC~~ *SUPER* POWERFUL

Getting started is easy:

1. Clone from [this repository](https://github.com/0x4F776C/ip)
2. Run `java -jar TrashBot.jar`
3. Start managing your tasks
4. Watch your productivity soar! ✨

Current Features:
- [x] Task management with todos
- [x] Event scheduling
- [x] Deadline tracking
- [ ] Data visualization
- [ ] Natural language processing
- [ ] Data archiving

For Java enthusiasts, here's the core command processing method:

```java
private Command parseCommand(String input) throws UnknownInputException {
    String[] parts = input.split(" ", 2);
    String commandType = parts[0].toLowerCase();

    switch (commandType) {
    case "todo":
        return new TodoCommand(input);
    case "deadline":
        return new DeadlineCommand(input);
    case "event":
        return new EventCommand(input);
    // ... more commands
    }
}
```

The `TrashBot` class serves as the central hub for all task management operations, handling everything from command parsing to file storage.

**Key Updates in this PR:**
* Added comprehensive JavaDoc comments
* Improved error handling
* Enhanced file storage system

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

0x4F776C committed:
Level-0 done

0x4F776C committed:
Level-1 done

0x4F776C committed:
Level-2 done

0x4F776C committed:
Level-3 done

0x4F776C committed:
Level-4 done

0x4F776C committed:
Level-5 done

0x4F776C committed:
Level-6 done

0x4F776C committed:
Replaced if-else with switch-case

0x4F776C committed:
Added Java Package for Features

0x4F776C committed:
Update iP progress - A-TextUiTesting

0x4F776C committed:
Level-7 done

0x4F776C committed:
Merge branch 'branch-Level-7'
iP Level-7 requirement

0x4F776C committed:
Added datetime implementation; W.I.P: Make the code more OOP style

0x4F776C committed:
Modified hierarchy to match OOP structure

0x4F776C committed:
Finalised Level-8

0x4F776C committed:
Merge branch 'branch-Level-8'
Level-8 Done

0x4F776C committed:
Renamed root package

0x4F776C committed:
Final Commit for A-Packages

0x4F776C committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Requirement for A-Gradle

0x4F776C committed:
A-Gradle and A-JUnit Done

0x4F776C committed:
Merge branch 'A-JUnit'
Requirement for A-Gradle and A-JUnit

0x4F776C committed:
Modified all instances of Duke/duke back to TrashBot/trashbot; A-Jar Done

0x4F776C committed:
Merge branch 'A-Jar'
Requirement for A-Jar

0x4F776C committed:
Removed StorageException class; Add JavaDoc to non-private class/method

0x4F776C committed:
Made changes to code based on Gradle::CheckStyleMain result

0x4F776C committed:
Add changes to preferences/settings based on SE-EDU recommendation

0x4F776C committed:
Level-9 Done

0x4F776C committed:
Merge branch 'branch-A-JavaDoc'

0x4F776C committed:
Fixed merge conflict with branch-A-CodingStandard

0x4F776C committed:
Fixed merge conflict with branch-Level-9

0x4F776C committed:
Ensured compliance with project requirements after resolving merge conflicts

0x4F776C committed:
Finalized necessary changes to align with project submission criteria

0x4F776C committed:
Pushed updates to meet project requirement and removed unnecessary files

0x4F776C committed:
Reorganize imports in commands package

The import statements in Command package do not follow the required
ordering specified by CheckStyleMain, which causes build validation
failures.

0x4F776C committed:
Reorganize imports in commands and core package

The import statements in command package and core package do not
follow the required ordering specified by CheckStyleMain,
which causes build validation failures.

0x4F776C committed:
Replace implicit import in TrashBot.java

The usage of implicit import is not recommended as it will import
all the classes from that directory. This could lead cluttered
namespace.

0x4F776C committed:
Fix missing HTML close tags in Javadoc comments

Javadoc comments contain unclosed HTML tags which cause parse errors
during documentation generation.

0x4F776C committed:
Reformat files to meet IntelliJ coding standards

Java files have inconsistent formatting that does
not align with IntelliJ's and SE-EDU code style settings.

0x4F776C committed:
Update gitignore list to include gradle files

0x4F776C committed:
Modify gitignore to previous status

0x4F776C committed:
Remove gradle related binaries and files from gitignore

0x4F776C committed:
Add ByeCommand to conform to OOP standard

0x4F776C committed:
Merge branch 'branch-A-CheckStyle'
Replace the prior "bye" feature with ByeCommand.java

0x4F776C committed:
Transform TrashBot to a fully UI experience

deannapoh commented on src/main/java/org/trashbot/storage/FileStorage.java:
```diff
@@ -0,0 +1,176 @@
+package org.trashbot.storage;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
```
I like how you listed the imported classes explicitly. 😃

deannapoh commented on src/main/java/org/trashbot/storage/FileStorage.java:
```diff
@@ -0,0 +1,176 @@
+     */
+    protected Task convertStringToTask(String str) {
+        try {
+            String[] parts = str.split(" \\| ", -1); // use "-1" to keep empty strings; OMG THIS WORKS
```
I like how you use trailing comments to explain your code. 👍

deannapoh commented on src/main/java/org/trashbot/tasks/Task.java:
```diff
@@ -0,0 +1,73 @@
+ */
+public class Task {
+    protected final String description;
+    protected boolean completed;
```
Perhaps you can use a more boolean-sounding name for boolean variables?

0x4F776C replied:
Thanks for pointing out! I totally missed this 🫠

deannapoh commented on src/main/java/org/trashbot/tasks/Task.java:
```diff
@@ -0,0 +1,73 @@
+     *
+     * @return true if task is done, false otherwise
+     */
+    public boolean isDone() {
```
I like how you named this boolean method to to sound like booleans. 😃

deannapoh commented on src/main/java/org/trashbot/core/TrashBot.java:
```diff
@@ -0,0 +1,123 @@
+ * </p>
+ */
+public class TrashBot {
+    private List<Task> tasks;
```
I like how you used plural form to represent a collection of objects. 💯

deannapoh commented on src/main/java/org/trashbot/commands/TodoCommand.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * The minimum string length for "Todo" filter
+     */
+    private static final int MIN_DESCRIPTION_LENGTH = 5;
```
I like how you capitalised constant names and used underscore to separate words. 🥇

deannapoh commented on src/main/java/org/trashbot/core/DataPersistence.java:
```diff
@@ -0,0 +1,61 @@
+
+import org.trashbot.tasks.Task;
+
+/**
+ * Defines the contract for persisting and retrieving task data in the task management system.
+ * This interface provides methods for saving tasks to and loading tasks from a persistent
+ * storage medium.
+ *
+ * <p>Implementations of this interface should handle all the necessary data serialization,
+ * deserialization, and storage operations. They should also manage any required resources
+ * (like file handles or database connections) appropriately.</p>
+ *
+ * <p>Example usage with a hypothetical implementation:
+ * <pre>
+ * DataPersistence storage = new FileDataPersistence("tasks.txt");
+ *
+ * // Saving tasks
+ * List&lt;Task&gt; tasks = new ArrayList<>();
+ * tasks.add(new Todo("Read book"));
+ * storage.save(tasks);
+ *
+ * // Loading tasks
+ * List&lt;Task&gt; loadedTasks = storage.load();
+ * </pre>
+ * </p>
+ *
+ * @see Task
+ */
```
I like how you wrote extremely descriptive comments for all public classes and methods.

deannapoh reviewed (commented):
Overall, your code looks good and is easy to read. 👍 
There were just a few minor coding standard violations regarding variables names. But otherwise, it looks good.

0x4F776C committed:
Modify commands to return String, and add JavaFx

0x4F776C committed:
Merge branch 'branch-Level-10'
Meet the requirement for Level-10

0x4F776C committed:
Remove trailing whitespace in MainWindow.java

0x4F776C committed:
Remove trailing whitespace in MainWindow.java

TheEnd-alr-taken commented on src/main/java/org/trashbot/exceptions/InvalidFormatException.java:
```diff
@@ -0,0 +1,17 @@
+ * This exception extends the {@link DukeException} class, enabling the creation of
+ * exception instances with a custom error message.
+ */
+public class InvalidFormatException extends DukeException {
```
I am curious. Is there any particular reason why you choose to separate DukeExceptions into the many custom exception specific to the bot app? I am having a hard time deciding as to whether I should deal with my exception more generically (Eg. all exceptions will just be DukeException(CustomErrorMessage) or specifically (in your case). What if there are too many catch exceptions in one method?

TheEnd-alr-taken commented on src/main/java/org/trashbot/tasks/Event.java:
```diff
@@ -0,0 +1,71 @@
+package org.trashbot.tasks;
+
+/**
```
I really like the effort you have place into creating all these detailed and informative javadoc comments.

TheEnd-alr-taken reviewed (commented):
Overall, looks good to me. I couldn't find anything else to comment on besides those that has already been pointed out by @deannapoh. Congrats on making it this far ahead.

0x4F776C committed:
Remove redundant comments

Because the comments were generating visual
clutter and this change improves readability by
reducing clutter while maintaining code clarity
through well-named variables and functions.

0x4F776C committed:
Remove redundant comments

Because the comments were generating visual
clutter and this change improves readability by
reducing clutter while maintaining code clarity
through well-named variables and functions.

0x4F776C committed:
Replace instances of explicit strings to variable

Because this might cause the need to modify each
instance of string individually. Hence, this
replacement can reduce the need to search line by
line.

0x4F776C committed:
Replace instances of explicit strings to variable

Because this might cause the need to modify each
instance of string individually. Hence, this
replacement can reduce the need to search line by
line.

0x4F776C committed:
Add assertion to check for empty or null input

Because this might cause the program to
not work as intended. Therefore, it is
crucial to check for the issues in a
programmer controlled environment.

# Conflicts:
#	src/main/java/org/trashbot/core/TrashBot.java

0x4F776C committed:
Add assertion to check for empty or null input

Because this might cause the program to
not work as intended. Therefore, it is
crucial to check for the issues in a
programmer controlled environment.

0x4F776C committed:
Refactor TrashBot.java code quality

Because the previous version is very
messy. Hence, I've decided to declutter
by extracting methods, implementing
SLAP, and adding ENUM to set constraint
to the Command types.

0x4F776C committed:
Refactor TrashBot.java code quality

Because the previous version is very
messy. Hence, I've decided to declutter
by extracting methods, implementing
SLAP, and adding ENUM to set constraint
to the Command types.

0x4F776C committed:
Merge pull request #1 from 0x4F776C/branch-A-Streams

Branch a streams

0x4F776C committed:
Merge branch 'master' into branch-A-Assertions

0x4F776C committed:
Merge pull request #2 from 0x4F776C/branch-A-Assertions

Add assertion to check for empty or null input

0x4F776C committed:
Merge branch 'master' into branch-A-CodeQuality

0x4F776C committed:
Merge pull request #3 from 0x4F776C/branch-A-CodeQuality

Refactor TrashBot.java code quality

0x4F776C committed:
Replace minor typos

Because the typos were misleading. Hence
removing them is the best choice.

0x4F776C committed:
Merge branch 'master'

Because there were conflicts
after resolving using web
method.

0x4F776C committed:
Add feature to mass delete tasks

Because it is tedious to delete
task one-by-one. Therefore, the
addition of this feature can
improve productivity.

0x4F776C committed:
Add CI to workflow

Because this will help to automate
my build, test, and deployment.

0x4F776C committed:
Add Java CI GitHub Actions workflow for automation

Currently, the project lacks a continuous integration (CI)
workflow, making it difficult to ensure code quality and
detect build issues early.

To address this, introduce a GitHub Actions workflow that
triggers on push and pull_request events. The workflow:

* Checks out the repository and ensures the correct branch is used.
* Validates the Gradle wrapper for security and integrity.
* Sets up JDK 17 with JavaFX support.
* Runs Gradle checks to verify build correctness.

This approach automates the build process, reduces manual
effort, and ensures consistency across different environments.
Using GitHub Actions allows seamless integration with the
repository, providing immediate feedback on code changes.

0x4F776C committed:
Refactor codebase for better code quality

Currently, the codebase contains TrashBotUI.java,
which is no longer used, leading to unnecessary clutter.
Additionally, FileStorage.java lacks sufficient structure,
making it harder to maintain. User input validation for
empty events and deadlines is also missing, reducing error
feedback quality.

To address this:

* Remove TrashBotUI.java as it serves no functional purpose.
* Refactor FileStorage.java using SLAP to improve maintainability.
* Implement checks for empty inputs to provide error prompting.

These changes improve code clarity, maintainability, and user
experience by ensuring only relevant code remains,
enhancing modularity, and making error messages more informative.

0x4F776C committed:
Modify GitHub action config

0x4F776C committed:
Populate README for documentation

0x4F776C committed:
Refine user documentation

0x4F776C committed:
Edit documentation to rectify redirection issue

0x4F776C committed:
Remove redirection issue within GFMD

0x4F776C committed:
Make minor change to test GitHub actions

0x4F776C committed:
Fix deadline bugs and update workflow

0x4F776C committed:
Add Action status and Ui to README

0x4F776C committed:
Modify docs README to include badge

0x4F776C committed:
Fix broken link issue with pages

0x4F776C committed:
Fix broken link issue with pages

0x4F776C committed:
Remove Ui.png from docs/readme

0x4F776C committed:
Remove unused exceptions

0x4F776C committed:
Fix JavaFx errors for *nix platforms
